sqlglot/dialects/postgres.py

```diff
@@ -60,16 +66,16 @@ def _date_diff_sql(self: Postgres.Generator, expression: exp.DateDiff) -> str:
     unit = expression.text("unit").upper()
     factor = DATE_DIFF_FACTOR.get(unit)
 
-    end = f"CAST({expression.this} AS TIMESTAMP)"
-    start = f"CAST({expression.expression} AS TIMESTAMP)"
+    end = f"CAST({self.sql(expression, 'this')} AS TIMESTAMP)"
+    start = f"CAST({self.sql(expression, 'expression')} AS TIMESTAMP)"
 
     if factor is not None:
         return f"CAST(EXTRACT(epoch FROM {end} - {start}){factor} AS BIGINT)"
 
     age = f"AGE({end}, {start})"
 
     if unit == "WEEK":
-        unit = f"EXTRACT(year FROM {age}) * 48 + EXTRACT(month FROM {age}) * 4 + EXTRACT(day FROM {age}) / 7"
+        unit = f"EXTRACT(days FROM ({end} - {start})) / 7"
```

Comment:
This must've been an oversight on my end when I implemented this logic... The fix was motivated by an [SO thread](https://stackoverflow.com/a/24075015), after I found out that the results were incorrect when transpiling a given test SQL query from Redshift to Postgres